balancer-js/src/modules/pools/pool-types/concerns/weighted/join.concern.ts

```diff
@@ -0,0 +1,189 @@
+    const sortedCalcInputs = this.sortCalcInputs(
+      parsedPoolInfo.tokens,
+      parsedPoolInfo.balances,
+      parsedPoolInfo.weights,
+      parsedPoolInfo.decimals,
+      tokensIn,
+      amountsIn,
+      wrappedNativeAsset
+    );
+    const expectedBPTOut = SDK.WeightedMath._calcBptOutGivenExactTokensIn(
+      sortedCalcInputs.balances.map((b) => new OldBigNumber(b)),
+      sortedCalcInputs.weights.map((w) => new OldBigNumber(w)),
+      sortedCalcInputs.amounts.map((a) => new OldBigNumber(a)),
+      new OldBigNumber(parsedPoolInfo.totalShares),
+      new OldBigNumber(parsedPoolInfo.swapFee)
+    ).toString();
+
+    const minBPTOut = subSlippage(
+      BigNumber.from(expectedBPTOut),
+      BigNumber.from(slippage)
+    ).toString();
+
+    const userData = WeightedPoolEncoder.joinExactTokensInForBPTOut(
+      sortedCalcInputs.amounts,
+      minBPTOut
+    );
+
+    const to = balancerVault;
+    const functionName = 'joinPool';
+    const attributes: JoinPool = {
+      poolId: pool.id,
+      sender: joiner,
+      recipient: joiner,
+      joinPoolRequest: {
+        assets: sortedCalcInputs.tokens,
+        maxAmountsIn: sortedCalcInputs.amounts,
+        userData,
+        fromInternalBalance: false,
+      },
+    };
+    const data = WeightedPoolJoin.encodeJoinPool(attributes);
+    const values = amountsIn.filter((amount, i) => tokensIn[i] === AddressZero); // filter native asset (e.g. ETH) amounts
+    const value = values[0] ? BigNumber.from(values[0]) : undefined;
+
+    return { to, functionName, attributes, data, value, minBPTOut };
+  }
+
+  // Helper methods
+
+  /**
+   * Sort BPT calc. inputs alphabetically by token addresses as required by calcBptOutGivenExactTokensIn
+   */
+  private sortCalcInputs = (
+    poolTokens: string[],
+    poolBalances: string[],
+    poolWeights: string[],
+    poolDecimals: number[],
+    tokensIn: string[],
+    amountsIn: string[],
+    wrappedNativeAsset: string
+  ) => {
+    const assetHelpers = new AssetHelpers(wrappedNativeAsset);
+    const [tokens, amounts] = assetHelpers.sortTokens(tokensIn, amountsIn) as [
+      string[],
+      string[]
+    ];
+    const [, balances, weights, decimals] = assetHelpers.sortTokens(
+      poolTokens,
+      poolBalances,
+      poolWeights,
+      poolDecimals
+    ) as [string[], string[], string[], number[]];
+    return {
+      tokens,
+      balances,
+      weights,
+      amounts,
+      decimals,
+    };
+  };
+
+  /**
+   * Parse pool info into EVM amounts
+   * @param {Pool}  pool
+   * @returns       parsed pool info
+   */
+  private parsePoolInfo = (pool: Pool) => {
+    const tokens = pool.tokens.map((token) => token.address);
+    const balances = pool.tokens.map((token) =>
+      parseFixed(token.balance, token.decimals).toString()
+    );
+    const weights = pool.tokens.map((token) => {
+      return parseUnits(token.weight || '0').toString(); // TODO: validate if null weights should indeed be considered zero
```

Comment:
We should probably throw if we're missing weights or decimals in a weighted pool.

Reply:
Done ✅